<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [1]:
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt
from matplotlib import gridspec

# import 'seaborn'
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import 'factorial' from math library
from math import factorial

# import 'stats' package from scipy library
from scipy import stats
from scipy.stats import randint
from scipy.stats import skewnorm

# import 'random' to generate a random sample
import random

# import statistics to perform statistical computation  
import statistics

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

# to test the normality 
import scipy.stats as st

# import the function to calculate the power of test
from statsmodels.stats import power

#### 1. From a sample of 250 observations, it is found that the average income of a 27 year old Londoner is £45,000 with a sample standard deviation of £4000. Obtain the 95% confidence interval to estimate the average income.

In [4]:
n =250
xbar = 45000
sigma = 4000
alpha = 1 - 0.95
zalphaby2 = st.norm.isf(alpha/2)
std_error = sigma/np.sqrt(n)
margin_error = zalphaby2 * std_error
lower_conf_limit = xbar - margin_error
upper_conf_limit = xbar + margin_error
print(lower_conf_limit, upper_conf_limit)

44504.16397415635 45495.83602584365


Hence we can say that population average income of londoner is lies between 44504 to 45495 at 95% configdence

In [6]:
st.norm.interval(1 - alpha, xbar, sigma/np.sqrt(n))

(44504.16397415635, 45495.83602584365)

In [8]:

st.norm.interval(0.99, xbar, sigma/np.sqrt(n)) # confidence level is 99%

(44348.361002958416, 45651.638997041584)

In [10]:
st.norm.interval(0.90, xbar, sigma/np.sqrt(n)) # confidence level is 90%

(44583.88128969955, 45416.11871030045)

#### 2. There are 150 apples on a tree. You randomly choose 40 apples and found that the average weight of apples is 182 grams with a standard deviation of 30 grams. Find the 95% confidence interval for the population mean.

In [11]:
n = 40
xbar = 182
sigma = 30
alpha = 1-0.95
std_error = sigma / np.sqrt(n)
dof = n -1
talphaby2 = st.t.isf(alpha / 2, dof)
margin_error = talphaby2 * std_error

In [12]:
lower_conf_limit = xbar - margin_error
upper_conf_limit = xbar + margin_error
print(lower_conf_limit, upper_conf_limit)

172.40553457459347 191.59446542540653


In [15]:
st.t.interval(0.95, dof, xbar, std_error) # Condifdence Level of 95%

(172.40553457459347, 191.59446542540653)

In [17]:
st.t.interval(0.9999, dof, xbar, std_error) # Condifdence Level of 99%

(161.44776610403363, 202.55223389596637)

In [20]:
alpha = 1 - 0.95
n = 10
print("Sample Size = ", n)
print("Zalphaby2 = ", st.norm.isf(alpha / 2))
print("Talphaby2 = ", st.t.isf(alpha / 2, n-1))

n = 20
print("Sample Size = ", n)
print("Zalphaby2 = ", st.norm.isf(alpha / 2))
print("Talphaby2 = ", st.t.isf(alpha / 2, n-1))

n = 100
print("Sample Size = ", n)
print("Zalphaby2 = ", st.norm.isf(alpha / 2))
print("Talphaby2 = ", st.t.isf(alpha / 2, n-1))

n = 1000
print("Sample Size = ", n)
print("Zalphaby2 = ", st.norm.isf(alpha / 2))
print("Talphaby2 = ", st.t.isf(alpha / 2, n-1))

Sample Size =  10
Zalphaby2 =  1.959963984540054
Talphaby2 =  2.2621571627409915
Sample Size =  20
Zalphaby2 =  1.959963984540054
Talphaby2 =  2.093024054408263
Sample Size =  100
Zalphaby2 =  1.959963984540054
Talphaby2 =  1.9842169515086827
Sample Size =  1000
Zalphaby2 =  1.959963984540054
Talphaby2 =  1.9623414611334487


#### 3. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Construct 99% confidence interval to estimate the population proportion ,Can you estimate the number of underperforming portfolios?

In [25]:
n = 13
X = 8
alpha = 1 - 0.99
p = X /n
q = 1-p
std_error = np.sqrt(p*q/n)
zalphaby2 = st.norm.isf(alpha/2)
margin_error = zalphaby2 * std_error
lower_conf_limit = p - margin_error
upper_conf_limit = p + margin_error
print(lower_conf_limit, upper_conf_limit) # range in percentage since its considered with probability
print(lower_conf_limit*50, upper_conf_limit*50) # to know the range out of 50 Portfolios


0.26782280814713805 0.9629464226220927
13.391140407356902 48.14732113110464


In [26]:
st.norm.interval(0.99, p, std_error)

(0.26782280814713805, 0.9629464226220927)

In [27]:
from statsmodels.stats.proportion import proportion_confint

proportion_confint(X, n, alpha)

(0.26782280814713805, 0.9629464226220927)

Hence we can estimate the underperforming portfolios out of 50 as 0.26*50 (13) to 0.96*50 (48) at 99% confidence level

In [28]:
df = sns.load_dataset('tips')
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

The Restaurent manager claims that the average total bill is $20, test this claim at 95% level of confidence or 5% of significance ?

In [29]:
# step 1: state  H0 & H1 and level of significance (alpha)

# H0: mu_tb = 20
# H1 mu_tb != 20
# significance is 5%
alpha = 0.05

##### Note: always consider the static data set as Sample of Population Data set irrespective of data set size
##### Population data set is always dynamic data set for example Realtime daat set , Time Series data

In [30]:
# Step 2: select appropiate test and compute corresponding test statistics
# one sample t test of mean (because of unknown population standard deviation), here standard deviation connot be calculated becuase we are having sample data set.
# t_statistic = (xbar - hypothesis(population) mean)/ std_error

mu = 20
n = len(df.total_bill)
xbar = df.total_bill.mean()
s = df.total_bill.std()
std_error = s / np.sqrt(n)
t_statistic = (xbar - mu) / std_error

print(t_statistic) # Statistical Value

-0.37559294451919506


In [31]:
talphaby2 = st.t.ppf(alpha/2, n - 1)
print(talphaby2) # Critical Region value

-1.9697743954258797


##### Critical Value Approach: (Decision Rule)
* If abs (test statistic) < abs(Critical value) > --> fail; to reject H0
    else reject H0
* since abs(-0.375) < abs(-1.96) ---> failed to reject H0

##### It means that the sample is drawn from a population whose mean is $20

#### **Conclusion: The Claim by the Restaurent Manager is True**

In [ ]:
st.t.interval(0.95, n-1, xbar, std_error)

##### Confidence Interval Approach: (Decision Rule)
* If hypothesis value falls in the confidence interval > --> fail; to reject H0
    else reject H0
* since 20 falls in between **18.66 and 20.90** ---> failed to reject H0

##### It means that the sample is drawn from a population whose mean is $20

In [34]:
p_value = st.t.cdf(t_statistic, n-1)*2
print(p_value)

0.7075471935626723


##### P - Value Approach: (Decision Rule)
* If p value is greater than alpha > --> fail; to reject H0
    else reject H0
* since p - value 0.70 > alpha(0.05) ---> failed to reject H0

##### It means that the sample is drawn from a population whose mean is $20

In [35]:
st.ttest_1samp(df.total_bill, mu)

TtestResult(statistic=-0.37559294451919506, pvalue=0.7075471935626723, df=243)